server_test.go

```diff
@@ -2,13 +2,204 @@ package veneur
 
 import (
 	"bytes"
+	"compress/zlib"
+	"encoding/json"
 	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 )
 
+// set up a boilerplate local config for later use
+func localConfig() Config {
+	return Config{
+		APIHostname:         "http://localhost",
+		Debug:               false,
+		Hostname:            "localhost",
+		Interval:            100 * time.Millisecond,
+		Key:                 "",
+		MetricMaxLength:     4096,
+		Percentiles:         []float64{.5, .75, .99},
+		ReadBufferSizeBytes: 2097152,
+		UDPAddr:             "localhost:8126",
+		HTTPAddr:            "localhost:8127",
+		ForwardAddr:         "http://localhost",
+		NumWorkers:          96,
+		NumReaders:          4,
+		StatsAddr:           "localhost:8125",
+		Tags:                []string{},
+		SentryDSN:           "",
+		FlushLimit:          1024,
+	}
+}
+
+// assertMetrics checks that all expected metrics are present
+// and have the correct value
+func assertMetrics(t *testing.T, metrics DDMetricsRequest, expectedMetrics map[string]float64) {
+	// it doesn't count as accidentally quadratic if it's intentional
+	for metricName, expectedValue := range expectedMetrics {
+		assertMetric(t, metrics, metricName, expectedValue)
+	}
+}
+
+func assertMetric(t *testing.T, metrics DDMetricsRequest, metricName string, value float64) {
+	defer func() {
+		if r := recover(); r != nil {
+			assert.Fail(t, "error extracting metrics", r)
+		}
+	}()
+	for _, metric := range metrics.Series {
+		if metric.Name == metricName {
+			assert.Equal(t, int(value+.5), int(metric.Value[0][1]+.5), metricName)
+			return
+		}
+	}
+	assert.Fail(t, "did not find expected metric", metricName)
+}
+
+// setupLocalServer creates a local server from the specified config
+// and starts listening for requests. It returns the server for inspection.
+func setupLocalServer(t *testing.T, config Config) Server {
```

Comment:
fair enough, let's refactor this into the server in a future pr to reduce the amount of code being duplicated here. we might also have to add an explicit panic handler field to the server object so that we can use `server.ConsumePanic` in prod and `t.Fatal` in tests.

also reiterating - could the flush loop cause spurious errors in the test? I think we might need a way to disable that part for this test (eg you add a bunch of metrics, invoke server.Flush manually at the end of the test and satisfy all the assertions, then in the background, the flush loop goroutine also invokes server.Flush a few millisec later, and since there are no metrics, it fails all the assertions)